ocaml/utils/language_extension_kernel_intf.mli

```diff
@@ -0,0 +1,71 @@
+(** These module types make it easier to import more source files directly
+    from the compiler into the [ppxlib_jane] open-source library. We copy source
+    files because the [ppxlib_jane] library can't rely on Jane Street's OCaml
+    compiler libraries -- it must remain compatible with the upstream OCaml
+    compiler.
+
+    The [Parsing.Jane_syntax_*] modules are implicitly functorized over which
+    module [Language_extension : Language_extension_kernel_for_jane_syntax] is
+    available. In the compiler, [Language_extension] refers to the module
+    defined in [utils/language_extension.ml], but in [ppxlib_jane],
+    [Language_extension] refers to a module that [ppxlib_jane] defines.
+*)
+
+module type Language_extension_kernel = sig
+  (** Language extensions provided by the Jane Street version of the OCaml
+      compiler.
+
+      This is the signature of the {!Language_extension_kernel} module that is
+      directly imported into [ppxlib_jane].
+  *)
+
+  type maturity = Stable | Beta | Alpha
+
+  (** The type of language extensions. An ['a t] is an extension that can either
+      be off or be set to have any value in ['a], so a [unit t] can be either on
+      or off, while a [maturity t] can have different maturity settings. *)
+  type _ t =
+    | Comprehensions : unit t
+    | Local : unit t
+    | Include_functor : unit t
+    | Polymorphic_parameters : unit t
+    | Immutable_arrays : unit t
+    | Module_strengthening : unit t
+    | Layouts : maturity t
+
+  module Exist : sig
+    type 'a extn = 'a t
+    type t = Pack : _ extn -> t
+
+    val all : t list
+  end with type 'a extn := 'a t
+
+  (** Print and parse language extensions; parsing is case-insensitive *)
+  val to_string : _ t -> string
+  val of_string : string -> Exist.t option
+  val maturity_to_string : maturity -> string
+
+  (** Check if a language extension is "erasable", i.e. whether it can be
+      harmlessly translated to attributes and compiled with the upstream
+      compiler. *)
+  val is_erasable : _ t -> bool
+end
+
+module type Language_extension_kernel_for_jane_syntax = sig
+  (** This module type defines the pieces of functionality used by
+      {!Jane_syntax_parsing} and {!Jane_syntax} so that we can more easily
+      import these modules into [ppxlib_jane], without also including all of the
+      [Language_extension] machinery.
+
+      In addition to the bindings from {!Language_extension_kernel}, it includes
+      the stateful operations that {!Jane_syntax_parsing} relies on. This limits
+      the number of bindings that [ppxlib_jane] needs to have mock
+      implementations for.
+  *)
+
+  include Language_extension_kernel
+
+  (** Check if a language extension is currently enabled. *)
+  val is_enabled : _ t -> bool
+  val is_at_least : 'a t -> 'a -> bool
```

Comment:
It seems to me like it would be nicer to *just* abstract over these two functions, and nothing else; we could share more code that way.

Reply:
I've given this a shot and like it more, thanks. Now, `Jane_syntax_parsing` is wired up to `Language_extensions_kernel` directly, and just these two bindings are pulled in from `Language_extension`.